cpp/src/arrow/array.h

```diff
@@ -60,8 +60,8 @@ class Array {
     return nulls_;
   }
 
-  bool Equals(const Array& arr) const;
-  virtual bool Equals(const std::shared_ptr<Array>& arr) const;
+  bool EqualsExact(const Array& arr) const;
+  virtual bool Equals(const std::shared_ptr<Array>& arr) const = 0;
```

Comment:
Clang revealed that this (plus the downstream PrimitiveArrayImpl templates) was causing some bad behavior, so we can review in the future how to have the cleanest code for the proliferation of array containers while avoiding unwanted template instantiation (I admit I don't know enough about this to make a judgment